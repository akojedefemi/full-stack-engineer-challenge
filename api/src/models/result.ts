import { Entity, Column, PrimaryGeneratedColumn, CreateDateColumn, UpdateDateColumn } from 'typeorm';
import { Length, IsDate, IsNotEmpty, IsOptional } from 'class-validator';

export enum ResultStatus {
    QUEUED = "Queued",
    IN_PROGRESS = "In Progress",
    SUCCESS = "Success",
    FAILURE = "Failure",
}

@Entity()
export class Result {
    @PrimaryGeneratedColumn()
    id: number;

    @Column({ type: 'enum', enum: ResultStatus, default: ResultStatus.QUEUED })
    status: string;

    @Column({ length: 80 })
    @Length(3, 80)
    @IsNotEmpty()
    repositoryName: string;

    @Column({ type: 'jsonb' })
    @IsNotEmpty()
    findings: object;

    @CreateDateColumn()
    queuedAt: Date;

    @Column({ nullable: true })
    @IsDate()
    @IsOptional()
    scanningAt: Date;

    @Column({ nullable: true })
    @IsDate()
    @IsOptional()
    finishedAt: Date;

    @UpdateDateColumn()
    updatedAt: Date;
}
export default Result;

// SCHEMA
const findingsExample = [{
    "type": "sast",
    "ruleId": "G402",
    "location": {
        "path": "connectors/apigateway.go",
        "positions": {
            "begin": {
                "line": 60
            }
        }
    },
    "metadata": {
        "description": "TLS InsecureSkipVerify set true.",
        "severity": "HIGH"
    }
}];
const exampleDate = new Date();

export const resultSchema = {
    id: { type: 'number', required: true, example: 1 },
    status: { type: 'string', required: false, example: 'Queued' },
    repositoryName: { type: 'string', required: true, example: 'my repository' },
    findings: { type: 'jsonb', required: true, example: findingsExample },
    queuedAt: { type: 'date', required: false, example: exampleDate },
    scanningAt: { type: 'date', required: false, example: exampleDate },
    finishedAt: { type: 'date', required: false, example: exampleDate },
    updatedAt: { type: 'date', required: false, example: exampleDate },
};