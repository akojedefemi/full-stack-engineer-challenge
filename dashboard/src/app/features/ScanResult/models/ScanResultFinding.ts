export interface ScanResultFinding {
  type: string;
  ruleId: string;
  location: Location;
  metadata: Metadata;
}

interface Location {
  path: string;
  positions: {
    begin: {
      line: number;
    }
  }
}

interface Metadata {
  severity: string;
  description: string;
}