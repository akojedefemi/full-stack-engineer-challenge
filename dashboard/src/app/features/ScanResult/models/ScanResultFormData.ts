export interface ScanResultFormData {
  repositoryName: string;
  findings: string;
}