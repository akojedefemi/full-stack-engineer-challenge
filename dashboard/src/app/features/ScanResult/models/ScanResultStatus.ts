export enum ScanResultStatus {
  QUEUED = "Queued",
  IN_PROGRESS = "In Progress",
  SUCCESS = "Success",
  FAILURE = "Failure",
}
