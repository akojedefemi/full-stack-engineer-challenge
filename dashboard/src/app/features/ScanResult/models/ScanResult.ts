import { ScanResultStatus } from "./ScanResultStatus";
import { ScanResultFinding } from "./ScanResultFinding";

export interface ScanResult {
  id: number;
  status: ScanResultStatus;
  repositoryName: string;
  findings: ScanResultFinding[];
  queuedAt: Date;
  scanningAt?: Date;
  finishedAt?: Date;
  updatedAt: Date;
}
