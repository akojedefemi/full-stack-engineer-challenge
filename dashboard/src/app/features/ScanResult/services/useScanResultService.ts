import { useEffect, useState } from 'react';

import { ScanResult } from '../models/ScanResult';
import { Service } from '../../Service';
import { normalizeResult } from './normalizeResult';
import { config } from '../../../config';

const useScanResultService = (id: string | number | undefined) => {
  const [result, setResult] = useState<Service<ScanResult>>({
    status: 'loading'
  });

  useEffect(() => {
    if (!id) {
      setResult({ status: 'error', error: new Error('not found') });
      return;
    }

    fetch(`${config.apiUrl}/api/v1/results/${id}`)
      .then(response => response.json())
      .then(normalizeResult)
      .then(response => setResult({ status: 'loaded', payload: response }))
      .catch(error => setResult({ status: 'error', error }));
  }, [id]);

  return result;
};

export default useScanResultService;