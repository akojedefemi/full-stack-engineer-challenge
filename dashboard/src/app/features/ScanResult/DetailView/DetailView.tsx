import React from 'react';
import { useParams } from 'react-router-dom';
import { Tree, Icon } from 'antd';

import { ScanResult } from '../models/ScanResult';
import useScanResultService from '../services/useScanResultService';
import Title from 'antd/lib/typography/Title';

const { TreeNode } = Tree;

const renderNode = (data: object, depth = '0') => {
  const keys = Object.keys(data);
  const values = Object.values(data);

  return values.map((value, i) => {
    const currentDepth = `${depth}-${i}`;

    if (value instanceof Date) {
      return <TreeNode icon={<Icon type="calendar" />} title={value.toLocaleString()} key={currentDepth} />;
    } else if (typeof value === 'number') {
      return <TreeNode icon={<Icon type="number" />} title={value} key={currentDepth} />;
    } else if (typeof value === 'object') {
      const title = depth === '0' ? `Finding #${keys[i]}` : keys[i];

      return (
        <TreeNode icon={<Icon type="smile-o" />} title={title} key={currentDepth}>
          {renderNode(value, currentDepth)}
        </TreeNode>
      );
    } else {
      return <TreeNode icon={<Icon type="barcode" />} title={value} key={currentDepth} />;
    }
  });
}

const renderDataView = (data: ScanResult) => {
  return (
    <>
      <Title>Findings of {data.repositoryName} ({data.status})</Title>
      <Tree
        showIcon={true}
        defaultExpandAll={true}
        switcherIcon={<Icon type="down" />}
      >
        {renderNode(data.findings)}
      </Tree>
    </>
  );
}

const DetailView: React.FC = () => {
  const { id } = useParams();
  const service = useScanResultService(id);

  return (
    <div>
      {service.status === 'loading' && <div>Loading...</div>}
      {service.status === 'loaded' && renderDataView(service.payload)}
      {service.status === 'error' && (
        <div>Error, there was an network problem. Please try again later.</div>
      )}
    </div>
  )
};

export default DetailView;
