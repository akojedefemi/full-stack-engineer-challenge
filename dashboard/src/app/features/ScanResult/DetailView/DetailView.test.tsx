import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

jest.mock('../services/useScanResultService');

import DetailView from './DetailView';
import { ScanResultStatus } from '../models/ScanResultStatus';
import useScanResultService from '../services/useScanResultService';

const useScanResultServiceMock = useScanResultService as jest.Mock<ReturnType<typeof useScanResultService>>;

const mockedScanResult = {
  id: 1,
  status: ScanResultStatus.QUEUED,
  repositoryName: 'someName',
  findings: [],
  queuedAt: new Date('2019-11-10T03:02:23.200Z'),
  scanningAt: undefined,
  finishedAt: undefined,
  updatedAt: new Date('2019-11-10T03:02:23.200Z'),
};

describe('DetailView', () => {
  beforeEach(() => jest.resetAllMocks());

  it('renders loading state', () => {
    useScanResultServiceMock.mockReturnValue({
      status: 'loading',
    });

    expect(renderer.create(<BrowserRouter><DetailView /></BrowserRouter>).toJSON()).toMatchSnapshot();
  });

  it('renders loaded state', () => {
    useScanResultServiceMock.mockReturnValue({
      status: 'loaded',
      payload: mockedScanResult,
    });

    expect(renderer.create(<BrowserRouter><DetailView /></BrowserRouter>).toJSON()).toMatchSnapshot();
  });

  it('renders error state', () => {
    useScanResultServiceMock.mockReturnValue({
      status: 'error',
      error: new Error('some error'),
    });

    expect(renderer.create(<BrowserRouter><DetailView /></BrowserRouter>).toJSON()).toMatchSnapshot();
  });

});
