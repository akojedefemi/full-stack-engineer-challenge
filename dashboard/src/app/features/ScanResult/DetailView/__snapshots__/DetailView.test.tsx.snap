// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailView renders error state 1`] = `
<div>
  <div>
    Error, there was an network problem. Please try again later.
  </div>
</div>
`;

exports[`DetailView renders loaded state 1`] = `
<div>
  <h1
    aria-label={null}
    className="ant-typography"
    style={
      Object {
        "WebkitLineClamp": null,
      }
    }
  >
    Findings of 
    someName
     (
    Queued
    )
  </h1>
  <ul
    className="ant-tree"
    role="tree"
    unselectable="on"
  />
</div>
`;

exports[`DetailView renders loading state 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
